File: ruby/bottle/bottle_with_comment.rb

class Bottles
  def song
    verses(99, 0)
  end

  def verses(hi, lo) # BAD: hi and lo are not specific enough
    hi.downto(lo).map {|n| verse(n) }.join("\n")
  end

  def verse(n) # BAD: n is too generic
    # BAD: inconsistent style
    # BAD: incomprehensibe code
    # BAD: a giant expression
    "#{n == 0 ? 'No more' : n} bottle#{'s' if n != 1}" + " of beer on the wall, " +
    "#{n == 0 ? 'no more' : n} bottle#{'s' if n != 1} of beer.\n" +
    "#{n > 0  ? "Take #{n > 1 ? 'one' : 'it'} down and pass it around" : "Go to the store and buy some more"}, " +
    "#{n-1 < 0 ? 99 : n-1 == 0 ? 'no more' : n-1} bottle#{'s' if n-1 != 1}" +
    " of beer on the wall.\n"
  end
end

File: php/bottle/BottleWithComment.php
<?php

class BottleWithComment
{
    public function song()
    {
        return $this->verses(99, 0);
    }

    public function verses(int $hi, int $lo) // BAD: hi and lo are not specific enough
    {
        $poem = '';
        for ( $n = $hi; $n >= $lo; $n--) {
            $poem .= $this->verse($n);
            $poem .= '</br>';
        }

        return $poem;
    }

    public function verse(int $n) // BAD: n is too generic
    {
        // BAD: inconsistent style
        // BAD: incomprehensible code
        // BAD: a giant expression
        $verse = '';

        $verse .=  $n == 0 ? 'No more bottles' :  ( $n == 1) ? '1 bottle' :  $n . ' bottles';
        $verse .= ' of beer on the wall, ';
        $verse .=  $n == 0 ? 'No more bottles' :  ( $n == 1) ? '1 bottle' :  $n . ' bottles' . ' of beer.';
        $verse .= '</br>';
        $verse .=  $n == 1 ? 'Go to the store and buy some more, ' : 'Take one down and pass it around, ';
        $verse .=  $n == 1 ? 'No more bottles' :  ( $n == 2) ? '1 bottle' :  $n == 0 ? '99 bottles' : ( $n - 1) . ' bottles';
        $verse .= ' of beer on the wall.';
        $verse .= '</br>';

        return $verse;
    }
}
